### PR TITLE
pthread cancellation support

### PR DESCRIPTION
**Pull request by Geertiebear**
This PR adds the following:

- Support for pthreads on linux
- ``pthread_setcancelstate``
- ``pthread_setcanceltype``
- ``pthread_cancel``
- ``pthread_testcancel``
- rtdl support for the ``GNU_PROPERTY`` PHDR

Currently there is no unwinding of the stack when we kill, which should be done in a future PR. Also we might want to set cancel types at points where we don't want to cancel inside the libc, but this is probably also for a future PR.

### CODE REVIEW

**Review by avdgrinten — APPROVED** [review]
Some comments below. Why do we move `x86_64-crt`to `crt-src/x86_64`?

File: options/internal/include/mlibc/tcb.hpp [hunk 1]
```diff
@@ -3,6 +3,26 @@
 #include <stdint.h>
 #include <bits/size_t.h>
 
+namespace {
+	// Set when the cancellation is enabled
+	constexpr unsigned int TCB_CANCELSTATE_BITMASK = 1 << 0;
```
[the remark this answers is not in the document]
Comment (Geertiebear) [reply options/internal/include/mlibc/tcb.hpp]: Done

File: options/internal/include/mlibc/tcb.hpp [hunk 2]
```diff
@@ -14,5 +34,6 @@ struct Tcb {
 	int didExit;
 	void *returnValue;
 	uintptr_t stackCanary;
+	int cancelBitmask;
```
[the remark this answers is not in the document]
Comment (Geertiebear) [reply options/internal/include/mlibc/tcb.hpp]: Done

File: options/internal/include/mlibc/tcb.hpp [hunk 3]
```diff
@@ -3,6 +3,26 @@
 #include <stdint.h>
 #include <bits/size_t.h>
 
+namespace {
+	// Set when the cancellation is enabled
+	constexpr unsigned int TCB_CANCELSTATE_BITMASK = 1 << 0;
+	// 1 - cancellation is asynchronous, 0 - cancellation is deferred
+	constexpr unsigned int TCB_CANCELTYPE_BITMASK = 1 << 1;
+	// Set when the thread has been cancelled
+	constexpr unsigned int TCB_CANCELLED_BITMASK = 1 << 1;
```
[the remark this answers is not in the document]
Comment (Geertiebear) [reply options/internal/include/mlibc/tcb.hpp]: No, that was fixed

File: options/internal/include/mlibc/tcb.hpp [hunk 4]
```diff
@@ -3,6 +3,26 @@
...
+
+namespace mlibc {
+	// Returns true when bitmask indicates thread has been asynchronously
+	// cancelled.
+	static constexpr bool tcb_async_cancelled(int value) {
+		return (value & (TCB_CANCELSTATE_BITMASK | TCB_CANCELTYPE_BITMASK
+				| TCB_CANCELLED_BITMASK)) == (TCB_CANCELSTATE_BITMASK
+					| TCB_CANCELTYPE_BITMASK | TCB_CANCELLED_BITMASK);
+	}
+}
```
[the remark this answers is not in the document]
Comment (Geertiebear) [reply options/internal/include/mlibc/tcb.hpp]: It checks that cancelling is enabled && cancelling is asynchronous && cancelling has been triggered.

File: options/posix/generic/pthread-stubs.cpp [hunk 5]
```diff
@@ -188,12 +189,84 @@ int pthread_getname_np(pthread_t, char *, size_t) {
...
-	void __do_cancel() {
-		//TODO(geert): stub for now
+
+	void sigcancel_handler(int signal, siginfo_t *info, void *ucontext) {
+		ucontext_t *uctx = static_cast<ucontext_t*>(ucontext);
+		// The function could be called from other signals, or from another
+		// process, in which case we should do nothing.
+		if (signal != SIGCANCEL || info->si_pid != getpid() ||
+				info->si_code != SI_TKILL)
+			return;
```
[the remark this answers is not in the document]
Comment (Geertiebear) [reply options/posix/generic/pthread-stubs.cpp:206]: It's for security reasons & a sanity check, e.g. a process from the outside could send a SIGCANCEL signal and shoot down a random thread.

**Comment by Geertiebear** [comment]
I moved the crt file to the arch subfolder just for consistency's sake.

**Review by avdgrinten — APPROVED** [review]
LGTM now, thanks a lot for the PR!

**Pull request merged**